[IMP] AWS load balancers do not provide X-Forwarded-Host.

amh-mw opened the pull request:
### Description of the issue/feature this PR addresses:

Amazon does not set the X-Forwarded-Host header for Classic and Application load balancers.

https://docs.aws.amazon.com/elasticloadbalancing/latest/application/x-forwarded-headers.html
https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/x-forwarded-headers.html

I have independently confirmed this with tcpdump and Wireshark.

### Current behavior before PR:

Even with --proxy-mode set, the request must also have the X-Forwarded-Host header set or the ProxyFix middleware is not activated.

### Desired behavior after PR is merged:

Odoo --proxy-mode works correctly for Amazon load balancers and any other reverse proxy that does not set X-Forwarded-Host.

I agree with the sentiment of the FIXME comment by @xmo-odoo. Checking for the presence of HTTP_X_FORWARDED_HOST is worse than not useful; it breaks intended functionality. We should rely upon Werkzeug to inspect the headers and adjust the environ appropriately.

https://werkzeug.palletsprojects.com/en/0.16.x/middleware/proxy_fix/

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/70117).

xmo-odoo commented:
@odony one more argument for removing the check. I don't think there's much of a security risk to it as the proxy mode remains an opt-in though I guess it's possible there are users who mis-set proxy-mode without a HOST set and are currently protected against their own misuse by the confluence of factors?

A secondary concern is that `HTTP_HOST` is used to set the `web.base.url`, without it being properly set URLs and redirections will be incorrectly generated, which can be difficult to diagnose and fix (cf #19848, #34412). Maybe we could add a warning in the log in case the `HTTP_HOST` is an IP and proxy mode is enabled, something along those lines? It's technically a valid setup but seems unlikely.

xmo-odoo commented:
> IIRC we kept this test to protect deployments from header injection. It's kind of tricky to deploy this right, as most proxies don't handle X-Forwarded headers the way we need it, by default.

We should probably update the comment I left back then in order to document that this was kept as a choice then.

amh-mw commented:
Since more recent versions of Odoo require `Werkzeug==0.16.1`, perhaps the user could be given  more granular control of ProxyFix arguments, i.e. set `x_host=0` via `--proxy-for=0` CLI argument?

If `web.base.url.freeze` were set, would that obviate the need for the safety fuse?

xmo-odoo commented:
> If `web.base.url.freeze` were set, would that obviate the need for the safety fuse?

Not really, that only avoids the `web.base.url` issue (which is limited in the first place because it's only updated when a system user logs in), but any code which relies on the `Host` header inside the application could get spoofed, which as odo notes critically includes the db filter, and would allow bypassing host-based filtering on database exposure.

amh-mw commented:
I have been following up with AWS support a bit and they pointed out that the load balancer populates the regular Host header with the value that I would have expected to be in X-Forwarded-Host. Still seems a bit odd to me, but workable?

xmo-odoo commented:
> I have been following up with AWS support a bit and they pointed out that the load balancer populates the regular Host header with the value that I would have expected to be in X-Forwarded-Host. Still seems a bit odd to me, but workable?

The issue is that Odoo has no way to know / be told whether it's in that situation or not as the feature is just an on/off switch. Providing a more granular configuration could solve that (basically tell the user to manage it), but it also requires additional complexity in that support for this has to be added in the configuration system at least (if not the CLI).

Although AWS setting Host and not even stripping X-Forwarded-Host makes things even weirder / worse, that means if proxy-mode is enabled the *client* (not the load balancer) controls whether it's used or not by sending X-Forwarded-Host or not. Why would they do that? It's weird.

amh-mw commented on odoo/service/wsgi_server.py:
```diff
@@ -108,17 +108,18 @@ def application_unproxied(environ, start_response):
     from werkzeug.middleware.proxy_fix import ProxyFix as ProxyFix_
     # 0.15 also supports port and prefix, but 0.14 only forwarded for, proto
     # and host so replicate that
-    ProxyFix = lambda app: ProxyFix_(app, x_for=1, x_proto=1, x_host=1)
```
I have been staring at this with my security hat on and noticed that Odoo sets x_host=1 here, whereas [Werkzeug defaults x_host=0](https://werkzeug.palletsprojects.com/en/0.16.x/middleware/proxy_fix/#werkzeug.middleware.proxy_fix.ProxyFix). It would be a breaking change, but we could address @odony 's concern by forcing users to opt into the dangerous behavior with `--proxy-x-host=1`.

amh-mw commented on odoo/tools/config.py:
```diff
@@ -137,9 +137,23 @@ def __init__(self, fname=None):
+        group.add_option("--proxy-x-proto", dest="proxy_x_proto", type="int", my_default=1,
+                         help="Number of values to trust for X-Forwarded-Proto.")
+        group.add_option("--proxy-x-host", dest="proxy_x_host", type="int", my_default=1,
+                         help="Number of values to trust for X-Forwarded-Host.")
```
Alternatively, the non-breaking change would be to allow users to opt out of the dangerous behavior with `--proxy-x-host=0`. This is what I am currently doing in my development environment.

amh-mw commented:
If the --proxy-x-* arguments made it into 15.0 and the breaking behavior enabled for 16.0, that would give users a year to learn about and preemptively set --proxy-x-host=1?

amh-mw commented:
Rebased again now that 16.0 is cut and we have a long runway again.

> Providing a more granular configuration could solve that (basically tell the user to manage it), but it also requires additional complexity in that support for this has to be added in the configuration system at least (if not the CLI).

@xmo-odoo Granular configuration implemented.

I went ahead and changed proxy_x_host=0 to be the default. Better a configuration problem during an upgrade to 17.0 than a security problem down the road.

Julien00859 commented:
The `test_misc1_reverse_proxy` test from the `test_http` test module is failing because the default configuration is to *not* trust the Host header set by the proxy. We can change the default value in the config or change the test to include `'proxy_x_host': 1` in the patch.

https://github.com/odoo/odoo/blob/b0ce0debd49493641f29cb75e43379b90d4302a0/odoo/addons/test_http/tests/test_misc.py#L45-L50

amh-mw commented:
@Julien00859 Thanks for the heads up on the broken build. I swear I used to get emails for them.

odony commented:
@amh-mw thanks for your efforts and follow-up!

Discussing this PR once more with @xmo-odoo, I've started to wonder about your deployment mode. Are you actually running pure Odoo services behind directly behind Amazon LB? That could explain the limited control you have over basic things like headers, etc.

Odoo was originally a web app deployed in a wsgi container. It's a bit confusing because `werkzeug` masquerades as a web server, but it's really not a production-ready web server.

In fact, our deployment documentation[^1] contains many sections that explain why it's necessary to run Odoo behind a proper web server/reverse proxy. It provides the missing bits:
- static assets service (e.g. not wasting an Odoo worker request for each static image)
- (as of Odoo 16) X-Accel/X-Sendfile service for static filestore contents
- gzip compression
- http/1.1, http/2, http/3 support
- rate-limiting and WAF-like protection
- ...

Is the AWS LB able to do all of the above, specifically the static file parts?

The lack of header control and specifically `X-Forwarded-Host` seems to be something we find in pure load-balancers (AWS LB, CloudFlare, ...). That makes sense because they aren't expected to alter the requests they're balancing, and the `Host` header is really part of the request, as opposed to the other XF headers.

From my experience, people are indeed running `nginx` or an equivalent server in front of Odoo, even if they sometimes have another LB in front of that. This may explain why we don't have many tickets/questions about the proxy mode. But I could be completely wrong here.


[^1]: https://www.odoo.com/documentation/latest/administration/install/deploy.html

Xavier-Do commented on odoo/tools/config.py:
```diff
@@ -142,13 +142,28 @@ def __init__(self, fname=None):
         group.add_option("--proxy-mode", dest="proxy_mode", action="store_true", my_default=False,
                          help="Activate reverse proxy WSGI wrappers (headers rewriting) "
                               "Only enable this when running behind a trusted web proxy!")
+        group.add_option("--proxy-x-for", dest="proxy_x_for", type="int", my_default=1,
+                         help="Number of values to trust for X-Forwarded-For.")
+        group.add_option("--proxy-x-proto", dest="proxy_x_proto", type="int", my_default=1,
+                         help="Number of values to trust for X-Forwarded-Proto.")
+        group.add_option("--proxy-x-host", dest="proxy_x_host", type="int", my_default=0,
+                         help="Number of values to trust for X-Forwarded-Host.")
+        group.add_option("--proxy-x-port", dest="proxy_x_port", type="int", my_default=0,
+                         help="Number of values to trust for X-Forwarded-Port.")
+        group.add_option("--proxy-x-prefix", dest="proxy_x_prefix", type="int", my_default=0,
+                         help="Number of values to trust for X-Forwarded-Prefix.")
+
```
I didn't really checked the issue here but in any case we prefer to avoid adding more command line arguments.

Xavier-Do replied:
Just god pinged here and didn't notice it was an old pr with an existing discussion, oupsie

Julien00859 commented:
`/web/assets` routes need to reach the Odoo process because Odoo needs to validate the access rights (same goes for `/web/content` and `/web/image`) but with X-Accel-Redirect enabled, nginx is still responsible to stream the file which is much better than having to waste one odoo worker to do it. See da8def8e410d and https://www.odoo.com/documentation/16.0/administration/install/deploy.html#serving-attachments

amh-mw commented:
I'm stuck on 15.0 until I have the time to finish migrating our custom modules to OWL, so unfortunately that comment doesn't apply to me (yet).

amh-mw commented:
It looks like AWS ALB have gained a setting since I created this issue:
<img width="476" alt="Screenshot 2023-08-23 at 10 19 54 AM" src="https://github.com/odoo/odoo/assets/495315/1e327055-3c06-49ee-8d60-c5c6f5daafb6">

I'll see if I can look into this in the near future and report back.

amh-mw commented:
I did finally look into the AWS ALB update. They add X-Forwarded-For by default now, but no movement on X-Forwarded-Host, so I am going to continue to run this patch on my containers, despite the issue being closed.
